packages/hash/shared/src/ProsemirrorManager.ts

```diff
@@ -0,0 +1,530 @@
+
+    if (typeof componentPosition !== "number") {
+      throw new Error(
+        "Cannot find inserted component node position in transaction",
+      );
+    }
+
+    return { tr, blockPosition, componentPosition };
+  }
+
+  /**
+   * This handles changing the block's data (blockEntity.properties.entity)
+   * to point to the targetEntity and updating the prosemirror tree to render
+   * the block with updated content
+   *
+   * @todo this does not work within text blocks
+   */
+  swapBlockData(
+    entityId: string,
+    targetChildEntity: EntityStoreType,
+    pos: number,
+  ) {
+    if (!this.view) {
+      throw new Error("Cannot trigger updateBlock without view");
+    }
+
+    const { tr } = this.view.state;
+    const entityStore = entityStorePluginStateFromTransaction(
+      tr,
+      this.view.state,
+    ).store;
+
+    const blockEntity = entityId ? entityStore.saved[entityId] : null;
+    const blockData = isBlockEntity(blockEntity)
+      ? blockEntity.properties.entity
+      : null;
+
+    if (!isBlockEntity(blockEntity) || !blockData) {
+      throw new Error("Can only update data of a BlockEntity");
+    }
+
+    // If the target entity is the same as the block's child entity
+    // we don't need to do anything
+    if (targetChildEntity.entityId === blockData.entityId) {
+      return;
+    }
+
+    const blockEntityDraftId = mustGetDraftEntityByEntityId(
+      entityStore.draft,
+      blockEntity.entityId,
+    ).draftId;
+
+    addEntityStoreAction(this.view.state, tr, {
+      type: "updateBlockEntityProperties",
+      payload: {
+        targetEntity: targetChildEntity,
+        blockEntityDraftId,
+      },
+    });
+
+    const updatedStore = entityStorePluginStateFromTransaction(
+      tr,
+      this.view.state,
+    ).store;
+
+    const newBlockNode = this.renderBlock(
+      blockEntity.properties.componentId,
+      updatedStore,
+      blockEntityDraftId,
+    );
+
+    tr.replaceRangeWith(pos, pos + newBlockNode.nodeSize, newBlockNode);
+    this.view.dispatch(tr);
+  }
+
+  /**
+   * Updates the provided properties on the specified entity.
+   * Merges provided properties in with existing properties.
+   * @param entityId the id of the entity to update
+   * @param propertiesToUpdate the properties to update
+   */
+  updateEntityProperties(entityId: string, propertiesToUpdate: JsonObject) {
+    if (!this.view) {
+      throw new Error("Cannot trigger updateEntityProperties without view");
+    }
+
+    const { tr } = this.view.state;
+
+    const entityStore = entityStorePluginStateFromTransaction(
+      tr,
+      this.view.state,
+    ).store.draft;
+
+    addEntityStoreAction(this.view.state, tr, {
+      type: "updateEntityProperties",
+      payload: {
+        draftId: mustGetDraftEntityByEntityId(entityStore, entityId).draftId,
+        properties: propertiesToUpdate,
+        merge: true,
+      },
+    });
+
+    this.view.dispatch(tr);
+  }
+
+  private async createNewDraftBlock(
```

Comment:
Maybe we could briefly explain what `draft` is for here? I don't think we do anywhere really

Reply:
I actually ended up renaming this in [`b663a10` (#872)](https://github.com/hashintel/hash/pull/872/commits/b663a10be8b24adc6c08577c1274f2f4d605ff68) to make it clearer what it does. Not sure this comment is relevant anymore – can you check?

Reply:
Yes, agreed, thank you